### PR TITLE
Add api endpoint to fetch resource raw yaml file of latest version

### PR DESCRIPTION
**Pull request by pratap0007**
This patch adds an endpoint to fetch resource raw yaml file for
latest version and also modify the existing api endpoint to fetch
resource raw yaml file for latest version and individual version

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/main/standards.md#principles) (if functionality changed/added)
- [x] Run API Unit Tests, Lint Checks, API Design, Golden Files with `make api-check`
- [ ] Run UI Unit Tests, Lint Checks with `make ui-check`
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/main/standards.md#commit-messages)

# Release Notes

```release-note
Add api endpoint to fetch resource raw yaml file of latest version
```

### CODE REVIEW

File: api/.env.dev [hunk 1]
```diff
@@ -2,7 +2,7 @@ ENVIRONMENT="development"
 
 POSTGRES_HOST="localhost"
 POSTGRES_PORT="5432"
-POSTGRES_DB="hub"
+POSTGRES_DB="test"
```
Comment (PuneetPunamiya) [inline api/.env.dev:5]: Do we need this change

File: api/v1/service/resource/resource.go [hunk 2]
```diff
@@ -508,13 +511,58 @@ func versionInfoFromVersion(v model.ResourceVersion) *resource.ResourceVersion {
...
+		}
+		p.Version = version
+	}
+
 	yamlPath := fmt.Sprintf("%s/%s/%s/%s/%s", s.CatalogClonePath(), strings.ToLower(p.Catalog), strings.ToLower(p.Kind), p.Name, p.Version)
 	yamlPath = fmt.Sprintf("%s/%s.yaml", yamlPath, p.Name)
 
 	content, err := os.ReadFile(yamlPath)
 	if err != nil {
+		s.Logger(ctx).Info("contekmwewww", err)
```
Comment (PuneetPunamiya) [inline api/v1/service/resource/resource.go:527]: Do we need this change

File: api/v1/service/resource/resource.go [hunk 3]
```diff
@@ -508,13 +511,58 @@ func versionInfoFromVersion(v model.ResourceVersion) *resource.ResourceVersion {
...
+		return nil, resource.MakeNotFound(fmt.Errorf("resource not found"))
+	}
+
+	return io.NopCloser(bytes.NewBuffer(content)), nil
+}
+
+// Fetch the latest verion of a resource
+func getLatestVersion(db *gorm.DB, name string, kind string) (string, error) {
+	var resInfo *model.Resource
+	if err := db.Where(&model.Resource{Name: name, Kind: strings.ToUpper(kind)}).Find(&resInfo).Error; err != nil {
```
Comment (PuneetPunamiya) [inline api/v1/service/resource/resource.go:558]: I think we should also query using catalog name as well

File: api/v1/service/resource/resource.go [hunk 4]
```diff
@@ -508,13 +511,58 @@ func versionInfoFromVersion(v model.ResourceVersion) *resource.ResourceVersion {
...
 	s.Logger(ctx).Info(fmt.Sprintf("Fetching YAML for resource %s", p.Name))
 
+	if p.Version == Latest {
+		version, err := getLatestVersion(s.DB(ctx), p.Name, p.Kind)
+		if err != nil {
+			return nil, err
+		}
+		p.Version = version
+	}
+
```
Comment (PuneetPunamiya) [inline api/v1/service/resource/resource.go:521]: I think there is no need to to modify this api and add latest in the api endpoint as we are already adding one more api to print the latest raw yaml

Comment (PuneetPunamiya) [reply api/v1/service/resource/resource.go:521]: cc @piyush-garg @vinamra28

File: api/v1/service/resource/resource.go (hunk 2 above)
Comment (pratap0007) [reply api/v1/service/resource/resource.go:527]: no, remove it

File: api/.env.dev (hunk 1 above)
Comment (pratap0007) [reply api/.env.dev:5]: No

File: api/v1/service/resource/resource.go [hunk 5]
```diff
@@ -28,12 +28,15 @@ import (
...
 	res "github.com/tektoncd/hub/api/pkg/shared/resource"
 	"github.com/tektoncd/hub/api/v1/gen/resource"
+	"gorm.io/gorm"
 )
 
 type service struct {
 	app.Service
 }
 
+const Latest = "latest"
```
Comment (vinamra28) [inline api/v1/service/resource/resource.go:38]: why are we exporting this variable? can we keep it as `const latest = "latest"` ?

File: api/v1/design/resource.go [hunk 6]
```diff
@@ -303,7 +303,34 @@ var _ = Service("resource", func() {
...
 		HTTP(func() {
 			GET("/resource/{catalog}/{kind}/{name}/{version}/raw")
+			Response(StatusOK)
+			Response("internal-error", StatusInternalServerError)
+			Response("not-found", StatusNotFound)
+			SkipResponseBodyEncodeDecode()
+		})
+	})
 
+	///
```
Comment (vinamra28) [inline api/v1/design/resource.go:313]: typo

File: api/v1/service/resource/resource.go [hunk 7]
```diff
@@ -518,3 +521,45 @@ func (s *service) GetRawYamlByCatalogKindNameVersion(ctx context.Context, p *res
...
+	}
+
+	var resInfo *model.Resource
+	if err := db.Where(&model.Resource{Name: name, Kind: strings.ToUpper(kind), CatalogID: catalogInfo.ID}).Find(&resInfo).Error; err != nil {
+		return "", resource.MakeInternalError(err)
+	}
+
+	var allVersions []string
+
+	if err := db.Model(&model.ResourceVersion{}).Select("version").Where(&model.ResourceVersion{ResourceID: resInfo.ID}).Find(&allVersions).Error; err != nil {
```
Comment (vinamra28) [inline api/v1/service/resource/resource.go:561]: noob question: will this return the versions in sorted order?

**Review by vinamra28 — COMMENTED** [review]
a couple of minor comments else PR looks good to me

File: api/v1/service/resource/resource.go (hunk 7 above)
Comment (pratap0007) [reply api/v1/service/resource/resource.go:561]: yes

**Review by vinamra28 — APPROVED** [review]

File: api/v1/service/resource/resource.go [hunk 8]
```diff
@@ -518,3 +519,45 @@ func (s *service) GetRawYamlByCatalogKindNameVersion(ctx context.Context, p *res
...
+		return "", resource.MakeInternalError(err)
+	}
+	return allVersions[len(allVersions)-1], nil
+}
```
Comment (PuneetPunamiya) [inline api/v1/service/resource/resource.go:563]: Can you please move this function to the shared/resource package i.e https://github.com/tektoncd/hub/blob/main/api/pkg/shared/resource/resource.go as we have all db queries in that package only and only the response related logic to be kept in this package and if possible could you please also add test for it

**Review by PuneetPunamiya — COMMENTED** [review]
Just one small nit to address
/approve

**Pull request merged**